KAFKA-14480 Move/Rewrite ProducerStateManager to storage module.

satishd opened the pull request:
KAFKA-14480 Move/Rewrite of ProducerStateManager to the storage module.

Replaced `File.listFiles` with `Files.list` in `ProducerStateManager.listSnapshotFiles`
Used Path instead of File in ProducerStateManager.isSnapshotFile to check whether the given path is a file or not and has a '.snpshot' suffix. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

satishd committed:
Moved ProducerStateManagerConfig to storage module.

satishd committed:
Moved ProducerStateManager to storage module

ijuma commented:
Thanks KAFKA-14478, I should have a PR soon.

satishd committed:
Added dynamic config changes for ProducerStateManagerConfig

satishd committed:
Cleanedup a few tests.

ijuma commented on core/src/main/scala/kafka/log/LogLoader.scala:
```diff
@@ -191,7 +192,7 @@ class LogLoader(
     // during log recovery may have deleted some files without the LogLoader.producerStateManager instance witnessing the
     // deletion.
-    producerStateManager.removeStraySnapshots(segments.baseOffsets.toSeq)
+    producerStateManager.removeStraySnapshots(segments.baseOffsets.toSeq.map(x => Long.box(x)).asJava)
```
Seems like we do two collection copies here. Maybe we can tweak things so that only one of them is needed (and when we convert the segments code, none are needed.

satishd replied:
Good catch! Avoided `Seq` conversion by using Collection as arg. We do not need these conversions when we move/rewrite LogSegment code, which is planned in few weeks.

ijuma commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -680,20 +680,20 @@ class UnifiedLog(@volatile var logStartOffset: Long,
+      (producerId.toLong, producerIdEntry.lastSeq)
     }
-  }
+  }.toMap
```
Can we avoid this `toMap` copy? Same for the case a few lines below.

satishd replied:
This is done to return an immutable `Map` from here.

ijuma replied:
I realized, but since we are creating a new map in this method, I don't see why an immutable Map is needed. Can we just change the result type to be `collection.Map`?

ijuma replied:
Resolving as I added more specific comments in the last review.

ijuma commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -2104,7 +2105,7 @@ object UnifiedLog extends Logging {
     if (recordVersion.value < RecordBatch.MAGIC_VALUE_V2 ||
-      (producerStateManager.latestSnapshotOffset.isEmpty && reloadFromCleanShutdown)) {
+      (producerStateManager.latestSnapshotOffset.asScala.isEmpty && reloadFromCleanShutdown)) {
       // To avoid an expensive scan through all of the segments, we take empty snapshots from the start of the
```
Can we call !isPresent instead of asScala.isEmpty?

satishd committed:
Addressed review comments

ijuma commented on core/src/test/scala/integration/kafka/api/TransactionsTest.scala:
```diff
@@ -652,7 +652,7 @@ class TransactionsTest extends IntegrationTestHarness {
 
       var producerStateEntry =
-        brokers(partitionLeader).logManager.getLog(new TopicPartition(testTopic, 0)).get.producerStateManager.activeProducers.head._2
+        brokers(partitionLeader).logManager.getLog(new TopicPartition(testTopic, 0)).get.producerStateManager.activeProducers.asScala.head._2
```
Can we use `.get(0)` instead of `asScala.head`? There's one other similar example in this file.

ijuma commented on core/src/test/scala/integration/kafka/api/TransactionsTest.scala:
```diff
@@ -685,7 +685,7 @@ class TransactionsTest extends IntegrationTestHarness {
       // check there that the epoch has actually increased
       producerStateEntry =
-        brokers(partitionLeader).logManager.getLog(new TopicPartition(testTopic, 0)).get.producerStateManager.activeProducers(producerId)
+        brokers(partitionLeader).logManager.getLog(new TopicPartition(testTopic, 0)).get.producerStateManager.activeProducers.get(producerId)
```
One difference in behavior is that Java's `Map.get` returns `null` if no mapping is found while Scala's `Map.apply` throws an exception. There's another similar example in this file.

satishd replied:
Added a not null check for `producerStateEntry`.

ijuma replied:
I don't see a not null check for `producerStateEntry`, am I missing something?

satishd replied:
aah, forgot to update the comment. I added that check initially, but later when I checked the old code, it did not have this kind of check in the old Scala code. I plan to add that as a followup, limiting this change mostly related to the refactoring.

ijuma replied:
It's fine for now since the test will still fail if no value is returned for the key, but it will result in a NPE instead of a NoSuchElementException.

ijuma commented on core/src/test/scala/unit/kafka/log/LogLoaderTest.scala:
```diff
@@ -557,7 +558,7 @@ class LogLoaderTest {
       keepPartitionMetadataFile = true)
 
-    verify(stateManager).removeStraySnapshots(any[Seq[Long]])
+    verify(stateManager).removeStraySnapshots((any[java.util.List[java.lang.Long]]))
```
Are the extra parenthesis around `any` needed? We have a few similar examples in this file.

ijuma commented on core/src/test/scala/unit/kafka/log/LogTestUtils.scala:
```diff
@@ -247,7 +246,7 @@ object LogTestUtils {
 
   def listProducerSnapshotOffsets(logDir: File): Seq[Long] =
-    ProducerStateManager.listSnapshotFiles(logDir).map(_.offset).sorted
+    ProducerStateManager.listSnapshotFiles(logDir).asScala.map(_.offset).sorted.toSeq
```
Is the `toSeq` at the end required? I'd have thought `sorted` returns a `Seq`.

satishd replied:
`sorted` returns a `Buffer` but not a `Seq`.

ijuma replied:
`Buffer` is a `mutable.Seq`, but this is probably returning `immutable.Seq`. Since it's a test, we can leave as is (I hadn't noticed it was a test in the initial review.

ijuma commented on core/src/test/scala/unit/kafka/log/UnifiedLogTest.scala:
```diff
@@ -215,11 +212,11 @@ class UnifiedLogTest {
     truncateFunc(reopened)(0L)
     assertEquals(None, reopened.firstUnstableOffset)
-    assertEquals(Map.empty, reopened.producerStateManager.activeProducers)
+    assertTrue(reopened.producerStateManager.activeProducers.isEmpty)
```
The previous approach is better as it includes the map contents in the message.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManagerConfig.java:
```diff
@@ -0,0 +1,37 @@
+        this.producerIdExpirationMs = producerIdExpirationMs;
+    }
+
+    public void updateProducerIdExpirationMs(int producerIdExpirationMs) {
```
Nit: use `set` instead of `update`?

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManagerConfig.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.Set;
+
+public class ProducerStateManagerConfig {
+    public static final Set<String> RECONFIGURABLE_CONFIGS = Collections.singleton("producer.id.expiration.ms");
```
We should have a static final for this config name here and reference it from `KafkaConfig`.

ijuma commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -1031,3 +1031,20 @@ class DynamicListenerConfig(server: KafkaBroker) extends BrokerReconfigurable wi
+
+  def validateReconfiguration(newConfig: KafkaConfig): Unit = {
+    if (newConfig.producerIdExpirationMs < 0)
+      throw new ConfigException(s"${KafkaConfig.ProducerIdExpirationMsProp} cannot be less than 0, current value is ${newConfig.producerIdExpirationMs}")
```
In the old code, the second parameter is `producerStateManagerConfig.producerIdExpirationMs`. I think we should actually have both the current and new value.

ijuma commented on core/src/test/scala/unit/kafka/log/ProducerStateManagerTest.scala:
```diff
@@ -88,7 +88,7 @@ class ProducerStateManagerTest {
     appendEndTxnMarker(stateManager, producerId, producerEpoch, ControlRecordType.COMMIT, offset = 27L)
 
-    val firstEntry = stateManager.lastEntry(producerId).getOrElse(throw new RuntimeException("Expected last entry to be defined"))
+    val firstEntry = stateManager.lastEntry(producerId).orElseThrow(() => throw new RuntimeException("Expected last entry to be defined"))
```
You are supposed to return the exception, not throw it (when using `orElseThrow`). Please update similar places in this file.

ijuma commented on core/src/test/scala/unit/kafka/log/ProducerStateManagerTest.scala:
```diff
@@ -192,7 +192,8 @@ class ProducerStateManagerTest {
     // should be able to append with the new epoch if we start at sequence 0
     append(stateManager, producerId, bumpedProducerEpoch, 0, 2L)
-    assertEquals(Some(0), stateManager.lastEntry(producerId).map(_.firstSeq))
+    val value: Optional[Int] = stateManager.lastEntry(producerId).map(_.firstSeq)
```
Do we need to extract the val here?

ijuma commented on core/src/test/scala/unit/kafka/log/ProducerStateManagerTest.scala:
```diff
@@ -340,8 +342,7 @@ class ProducerStateManagerTest {
       producerStateManagerConfig, time)
-    reloadedStateManager.truncateAndReload(logStartOffset = 0L,
-      logEndOffset = stateManager.mapEndOffset, currentTimeMs = time.milliseconds())
+    reloadedStateManager.truncateAndReload(0L,stateManager.mapEndOffset, time.milliseconds())
```
Nit: space missing after `,`.

ijuma reviewed (commented):
Thanks for the updates. I completed all the files except `ProducerStateManager`. See below.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+ * been deleted.
+ */
+public class ProducerStateManager {
+    private static final Logger log = LoggerFactory.getLogger(ProducerStateManager.class.getName());
```
Nit: you can pass the class, no need to get `getName()`.

ijuma replied:
Also, we have a `logIdent` defined for the class logger, so we probably need a non static logger.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+
+    // Remove these once UnifiedLog moves to storage module.
+    public static final String DELETED_FILE_SUFFIX = ".deleted";
+    public static final String PRODUCER_SNAPSHOT_FILE_SUFFIX = ".snapshot";
```
We have `LogFileUtils` for these kinds of things.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+    private static final int PRODUCER_ENTRIES_OFFSET = CRC_OFFSET + 4;
+
+    private static final Schema PRODUCER_SNAPSHOT_ENTRY_SCHEMA = new Schema(new Field(PRODUCER_ID_FIELD, Type.INT64, "The producer ID"), new Field(PRODUCER_EPOCH_FIELD, Type.INT16, "Current epoch of the producer"), new Field(LAST_SEQUENCE_FIELD, Type.INT32, "Last written sequence of the producer"), new Field(LAST_OFFSET_FIELD, Type.INT64, "Last written offset of the producer"), new Field(OFFSET_D ...
+    private static final Schema PID_SNAPSHOT_MAP_SCHEMA = new Schema(new Field(VERSION_FIELD, Type.INT16, "Version of the snapshot file"), new Field(CRC_FIELD, Type.UNSIGNED_INT32, "CRC of the snapshot data"), new Field(PRODUCER_ENTRIES_FIELD, new ArrayOf(PRODUCER_SNAPSHOT_ENTRY_SCHEMA), "The entries in the producer table"));
```
The formatting of this is odd. The formatting we used in the Scala code looks more readable to me.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+            long crc = struct.getUnsignedInt(CRC_FIELD);
+            long computedCrc = Crc32C.compute(buffer, PRODUCER_ENTRIES_OFFSET, buffer.length - PRODUCER_ENTRIES_OFFSET);
+            if (crc != computedCrc)
+                throw new CorruptSnapshotException("Snapshot is corrupt (CRC is no longer valid). " + "Stored crc: $crc. Computed crc: $computedCrc");
```
This needs updating not to use string interpolation as it doesn't work in Java.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+                throw new CorruptSnapshotException("Snapshot is corrupt (CRC is no longer valid). " + "Stored crc: $crc. Computed crc: $computedCrc");
+
+            List<ProducerStateEntry> entries = new ArrayList<>();
+            for (Object producerEntryObj : struct.getArray(PRODUCER_ENTRIES_FIELD)) {
```
We can assign the array to a variable and then allocate `entries` with the right size.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+                Optional<BatchMetadata> batchMetadata =
+                        (offset >= 0) ? Optional.of(new BatchMetadata(seq, offset, offsetDelta, timestamp)) : Optional.empty();
+
+                entries.add(new ProducerStateEntry(producerId, producerEpoch, coordinatorEpoch, timestamp, (currentTxnFirstOffset >= 0) ? OptionalLong.of(currentTxnFirstOffset) : OptionalLong.empty(), batchMetadata));
```
I think the old code where `currentTxnFirstOffsetValue` was a separate variable was more readable.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+    // visible for testing
+    public static List<SnapshotFile> listSnapshotFiles(File dir) throws IOException {
+        if (dir.exists() && dir.isDirectory()) {
+            try (Stream<Path> paths = Files.list(dir.toPath())) {
```
Worth noting that we changed the implementation to use `Files.list`. This is an improvement, but we should call out these changes in the PR description.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+    public static List<SnapshotFile> listSnapshotFiles(File dir) throws IOException {
+        if (dir.exists() && dir.isDirectory()) {
+            try (Stream<Path> paths = Files.list(dir.toPath())) {
+                return paths.filter(path -> path.toFile().isFile() && isSnapshotFile(path.toFile())).map(path -> new SnapshotFile(path.toFile())).collect(Collectors.toList());
```
It's inefficient to make multiple `toFile` calls for each file. If we want to keep this improved implementation, we should change `isSnapshotFile` to work with `Path` and use the relevant method to check that a `Path` is a file.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+            Struct producerEntryStruct = struct.instance(PRODUCER_ENTRIES_FIELD);
+            producerEntryStruct.set(PRODUCER_ID_FIELD, producerId)
+                    .set(PRODUCER_EPOCH_FIELD, entry.producerEpoch())
+                    .set(LAST_SEQUENCE_FIELD, entry.lastSeq()).set(LAST_OFFSET_FIELD, entry.lastDataOffset())
```
We should have a single `set` per line.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+        Struct struct = new Struct(PID_SNAPSHOT_MAP_SCHEMA);
+        struct.set(VERSION_FIELD, PRODUCER_SNAPSHOT_VERSION);
+        struct.set(CRC_FIELD, 0L); // we'll fill this after writing the entries
+        List<Struct> structEntries = new ArrayList<>(entries.size());
```
We can avoid a copy by using Struct[] here. Since we know the size, the code will be similar, the main difference is having to keep an index that we increment inline as we add an item to the array (e.g. `structEntries[i++]  = producerEntryStruct`.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+    private static final Schema PRODUCER_SNAPSHOT_ENTRY_SCHEMA = new Schema(new Field(PRODUCER_ID_FIELD, Type.INT64, "The producer ID"), new Field(PRODUCER_EPOCH_FIELD, Type.INT16, "Current epoch of the producer"), new Field(LAST_SEQUENCE_FIELD, Type.INT32, "Last written sequence of the producer"), new Field(LAST_OFFSET_FIELD, Type.INT64, "Last written offset of the producer"), new Field(OFFSET_D ...
+    private static final Schema PID_SNAPSHOT_MAP_SCHEMA = new Schema(new Field(VERSION_FIELD, Type.INT16, "Version of the snapshot file"), new Field(CRC_FIELD, Type.UNSIGNED_INT32, "CRC of the snapshot data"), new Field(PRODUCER_ENTRIES_FIELD, new ArrayOf(PRODUCER_SNAPSHOT_ENTRY_SCHEMA), "The entries in the producer table"));
+    private final TopicPartition topicPartition;
+    private volatile File logDir;
```
Let's group the fields based on whether they're `final`, `volatile`, mutable, etc.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+        return producerStateManagerConfig;
+    }
+
+    public boolean hasLateTransaction(long currentTimeMs) {
```
We should add a javadoc that explains that this method is thread safe (the Scala class had an annotation indicating that).

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+
+
+    /**
+     * Load producer state snapshots by scanning the _logDir.
```
`_logDir` doesn't exist in the Java class.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+    public Optional<LogOffsetMetadata> firstUnstableOffset() {
+        Optional<LogOffsetMetadata> unreplicatedFirstOffset = Optional.ofNullable(unreplicatedTxns.firstEntry()).map(e -> e.getValue().firstOffset);
+        Optional<LogOffsetMetadata> undecidedFirstOffset = Optional.ofNullable(ongoingTxns.firstEntry()).map(e -> e.getValue().firstOffset);
+        Optional<LogOffsetMetadata> result;
```
Why do we need this? Seems like we can inline it by using `return` within the if/else.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+    /**
+     * Get a copy of the active producers
+     */
+    public Map<Long, ProducerStateEntry> activeProducers() {
```
The documentation says a copy is returned, but the implementation does not do that.

ijuma replied:
Looks like this wasn't addressed, but I added a new comment related to it in the latest review, so resolving this one.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+            if (latestSnapshotFileOptional.isPresent()) {
+                SnapshotFile snapshot = latestSnapshotFileOptional.get();
+                try {
+                    log.info("Loading producer state from snapshot file '{}'", snapshot.file());
```
The original code used `snapshot` instead of `snapshot.file()` for the log parameter.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+                    updateOldestTxnTimestamp();
+                    return;
+                } catch (CorruptSnapshotException e) {
+                    log.warn("Failed to load producer snapshot from '${snapshot.file}': ${e.getMessage}");
```
Replace string interpolation with log interpolation.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+        }
+    }
+
+    public void loadProducerEntry(ProducerStateEntry entry) {
```
The old code had a `visible for testing` comment. Is that not true?

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+     * Expire any producer ids which have been idle longer than the configured maximum expiration timeout.
+     */
+    public void removeExpiredProducers(long currentTimeMs) {
+        Set<Long> keys = producers.entrySet().stream().filter(entry -> isProducerExpired(currentTimeMs, entry.getValue())).map(Map.Entry::getKey).collect(Collectors.toSet());
```
Could this be a `List` instead of `Set`? Not sure we need the latter here.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+    public void removeExpiredProducers(long currentTimeMs) {
+        Set<Long> keys = producers.entrySet().stream().filter(entry -> isProducerExpired(currentTimeMs, entry.getValue())).map(Map.Entry::getKey).collect(Collectors.toSet());
+        for (Long key : keys) {
+            producers.remove(key);
```
`producers.keySet().removeAll()` is probably more efficient.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+     */
+    public void update(ProducerAppendInfo appendInfo) {
+        if (appendInfo.producerId() == RecordBatch.NO_PRODUCER_ID)
+            throw new IllegalArgumentException("Invalid producer id " + appendInfo.producerId() + " passed to update " + "for partition $topicPartition");
```
Replace string interpolation - doesn't work in Java.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+        if (appendInfo.producerId() == RecordBatch.NO_PRODUCER_ID)
+            throw new IllegalArgumentException("Invalid producer id " + appendInfo.producerId() + " passed to update " + "for partition $topicPartition");
+
+        log.trace("Updated producer " + appendInfo.producerId() + " state to" + appendInfo);
```
Use log interpolation.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+
+        log.trace("Updated producer " + appendInfo.producerId() + " state to" + appendInfo);
+        ProducerStateEntry updatedEntry = appendInfo.toEntry();
+        ProducerStateEntry producerStateEntry = producers.get(appendInfo.producerId());
```
Shall we call this `currentEntry` like in the old code?

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+        }
+    }
+
+    private File producerSnapshotFile(File logDir, long offset) {
```
This and the next private method should be in `LogFileUtils`?

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+    public void completeTxn(CompletedTxn completedTxn) {
+        TxnMetadata txnMetadata = ongoingTxns.remove(completedTxn.firstOffset);
+        if (txnMetadata == null)
+            throw new IllegalArgumentException("Attempted to complete transaction $completedTxn on partition $topicPartition " + "which was not started");
```
String interpolation must be fixed.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+        updateOldestTxnTimestamp();
+    }
+
+    public void deleteSnapshotsBefore(long offset) throws IOException {
```
We should document that the method is thread-safe (like the old class).

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+     * <p>
+     * Note: This method is safe to use with async deletes. If a race occurs and the snapshot file
+     * is deleted without this ProducerStateManager instance knowing, the resulting exception on
+     * SnapshotFile rename will be ignored and None will be returned.
```
Comment must be updated not to refer to `None`.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+     * is deleted without this ProducerStateManager instance knowing, the resulting exception on
+     * SnapshotFile rename will be ignored and None will be returned.
+     */
+    public Optional<SnapshotFile> removeAndMarkSnapshotForDeletion(long snapshotOffset) throws IOException {
```
Did we make this public due to the callers being in different files (previously same package, but no longer)?

satishd replied:
Right.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+                snapshotFile.renameTo(DELETED_FILE_SUFFIX);
+                return Optional.of(snapshotFile);
+            } catch (NoSuchFileException ex) {
+                log.info("Failed to rename producer state snapshot ${snapshot.file.getAbsoluteFile} with deletion suffix because it was already deleted");
```
Use log interpolation, not string interpolation.

ijuma reviewed (commented):
Completed the review. It's probably a good idea for you to take an initial review pass before asking for other reviews as there were a number of cases where things weren't converted properly (especially when dealing with strings).

satishd commented:
@ijuma It looks like the changes that were done in the conversions were accidentally dropped in my local repo while doing a few rebases in my trunk. I should have checked that before pushing to PR and pinging for review. Sorry about that.

satishd committed:
Addressed review comments

satishd commented:
@ijuma It looks like the changes that were done in the conversions were accidentally dropped in my local repo while doing a few rebases and squashes with trunk and other changes. I should have checked that before pushing to PR and pinging for review, which I generally follow was missed. My sincere apologies on that.

ijuma commented:
@satishd No worries, it happens to all of us. :)

Hangleton commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+    public void removeStraySnapshots(Collection<Long> segmentBaseOffsets) throws IOException {
+        OptionalLong maxSegmentBaseOffset = segmentBaseOffsets.isEmpty() ? OptionalLong.empty() : OptionalLong.of(segmentBaseOffsets.stream().max(Long::compare).get());
+
+        HashSet<Long> baseOffsets = new HashSet<>(segmentBaseOffsets);
```
`Set<Long> baseOffsets`?

Hangleton commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManagerConfig.java:
```diff
@@ -0,0 +1,37 @@
+    public int producerIdExpirationMs() {
+        return producerIdExpirationMs;
+    }
+}
```
Missing CR.

Hangleton commented on storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java:
```diff
@@ -0,0 +1,671 @@
+    /**
+     * Take a snapshot at the current end offset if one does not already exist.
+     */
+    public void takeSnapshot() throws IOException {
```
Curious - is this exception eventually going to be exposed as an `UncheckedIOException` ([KAFKA-14490](https://issues.apache.org/jira/browse/KAFKA-14490))? 

Is there a reason not to invoke a log failure handler (because we want to propagate to the log layer?)?

ijuma replied:
We are trying to avoid making changes that could result in subtle breakage unless there is a strong reason to make the change and we have a high degree of confidence that it's safe to do so. With that in mind, what's the reason for making the change you are suggesting?

Hangleton replied:
Thanks for the follow-up. I missed to mention the question above is beyond the scope of this PR. I am trying to understand the strategy adopted in case of I/O failures in the data path. Some I/O failures directly results in putting the log directory written on offline. Some others are propagated up the stack. Is there a design pattern applying here that we have to follow? Do all I/O failures warrant the log directory to be made offline?

Hangleton replied:
Apologies - I just found an answer to this question [here](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/errors/KafkaStorageException.java#L23-L28).

Quoting from `KafkaStorageException`'s Javadoc:

> 1) If the server has not finished loading logs, IOException does not need to be converted to KafkaStorageException
> 2) After the server has finished loading logs, IOException should be caught and trigger LogDirFailureChannel.maybeAddOfflineLogDir(). Then the IOException should either be swallowed and logged, or be converted and re-thrown as KafkaStorageException
> 3) It is preferred for IOException to be caught in Log rather than in ReplicaManager or LogSegment.

Skimmed through the code paths in the server codebase, and I/O exceptions thrown from here are propagated up to `UnifiedLog` and log directories added there to the offline list, following the guidance from 3).

satishd committed:
Addressed review comments

satishd commented:
Thanks @ijuma for the review. Addressed them inline and/or with the latest commits.

ijuma commented:
Thanks for the updates, I'll review them in a couple of days.

showuon commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -0,0 +1,679 @@
+        snapshots = loadSnapshots();
+    }
+
+
```
nit: additional empty line

showuon commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -0,0 +1,679 @@
+    public void onHighWatermarkUpdated(long highWatermark) {
+        removeUnreplicatedTransactions(highWatermark);
+    }
+
```
ditto

showuon commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -0,0 +1,679 @@
+        ProducerStateEntry currentEntry = lastEntry(producerId).orElse(ProducerStateEntry.empty(producerId));
+        return new ProducerAppendInfo(topicPartition, producerId, currentEntry, origin);
+    }
+
```
ditto

showuon commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -0,0 +1,679 @@
+
+        updateOldestTxnTimestamp();
+    }
+
```
ditto

showuon commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -0,0 +1,679 @@
+            if (lastOffset.isPresent() && lastOffset.getAsLong() < offset) iterator.remove();
+        }
+    }
+
```
ditto

showuon commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -0,0 +1,679 @@
+     * transaction index, but the completion must be done only after successfully appending to the index.
+     */
+    public long lastStableOffset(CompletedTxn completedTxn) {
+        return findNextIncompleteTxn(completedTxn.producerId).map(x -> x.firstOffset.messageOffset).orElse(completedTxn.lastOffset + 1);
```
nit: could we make them 3 lines for each method?
return findNextIncompleteTxn(completedTxn.producerId)
  .map(x -> x.firstOffset.messageOffset)
  .orElse(completedTxn.lastOffset + 1);

showuon reviewed (approved):
LGTM! Just some minor comments. Thanks.

satishd committed:
Addressed review comments

satishd commented:
Thanks @showuon for your review. Addressed them with the latest commit.

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -0,0 +1,676 @@
+        this.maxTransactionTimeoutMs = maxTransactionTimeoutMs;
+        this.producerStateManagerConfig = producerStateManagerConfig;
+        this.time = time;
+        log = new LogContext("[ProducerStateManager partition=" + topicPartition + ")").logger(ProducerStateManager.class);
```
Closing bracket should be `]`, not `)`.

ijuma commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -45,7 +45,7 @@ import org.apache.kafka.server.record.BrokerCompressionType
-import org.apache.kafka.storage.internals.log.{AbortedTxn, AppendOrigin, BatchMetadata, CompletedTxn, EpochEntry, FetchDataInfo, FetchIsolation, LastRecord, LogConfig, LogDirFailureChannel, LogOffsetMetadata, LogValidator, ProducerAppendInfo}
+import org.apache.kafka.storage.internals.log.{AbortedTxn, AppendOrigin, BatchMetadata, CompletedTxn, EpochEntry, FetchDataInfo, FetchIsolation, LastRecord, LogConfig, LogDirFailureChannel, LogOffsetMetadata, LogValidator, ProducerAppendInfo, ProducerStateManager, ProducerStateManagerConfig}
 
 import scala.annotation.nowarn
```
We should remove things like `ProducerSnapshotFileSuffix` and `producerSnapshotFile` since they're now in `LogFileUtils`.

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -0,0 +1,676 @@
+            long crc = struct.getUnsignedInt(CRC_FIELD);
+            long computedCrc = Crc32C.compute(buffer, PRODUCER_ENTRIES_OFFSET, buffer.length - PRODUCER_ENTRIES_OFFSET);
+            if (crc != computedCrc)
+                throw new CorruptSnapshotException("Snapshot is corrupt (CRC is no longer valid). " + "Stored crc: " + crc + ". Computed crc: " + computedCrc);
```
The first `+` seems unnecessary.

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -0,0 +1,676 @@
+     * Get a copy of the active producers
+     */
+    public Map<Long, ProducerStateEntry> activeProducers() {
+        return Collections.unmodifiableMap(producers);
```
This is not a copy unlike what the documentation says. If we don't actually need to copy, we should change the documentation. Looking at the code, it does seem like we do _not_ need a copy, so we probably just need to update the documentation.

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -0,0 +1,676 @@
+     */
+    public void update(ProducerAppendInfo appendInfo) {
+        if (appendInfo.producerId() == RecordBatch.NO_PRODUCER_ID)
+            throw new IllegalArgumentException("Invalid producer id " + appendInfo.producerId() + " passed to update " + "for partition" + topicPartition);
```
Unnecessary `+`  between `to update` and `for partition`.

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -0,0 +1,676 @@
+    public void completeTxn(CompletedTxn completedTxn) {
+        TxnMetadata txnMetadata = ongoingTxns.remove(completedTxn.firstOffset);
+        if (txnMetadata == null)
+            throw new IllegalArgumentException("Attempted to complete transaction " + completedTxn + " on partition " + topicPartition + "which was not started");
```
Space missing before `which`.

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -0,0 +1,676 @@
+    }
+
+    /**
+     * Deletes the producer snapshot files until the given offset(exclusive) in a thread safe manner.
```
Nit: there should be a space before `(exlusive)`.

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -0,0 +1,676 @@
+     * Deletes the producer snapshot files until the given offset(exclusive) in a thread safe manner.
+     *
+     * @param offset offset number
+     * @throws IOException if any IOException while deleting the files.
```
Looks like a word is missing after `if any IOException`?

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -0,0 +1,676 @@
+
+    private static boolean isSnapshotFile(Path path) {
+        File file = path.toFile();
+        return file.isFile() && file.getName().endsWith(LogFileUtils.PRODUCER_SNAPSHOT_FILE_SUFFIX);
```
I think we can implement this directly on the path without converting to File.

satishd replied:
I am not sure there is a way to find whether a given path instance is a `file` or not like the way we check `File.isFile()`.

ijuma replied:
How about `Files.isRegularFile(path)`?

satishd replied:
Thanks for that. I will make the change and update the PR.

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogFileUtils.java:
```diff
@@ -32,4 +37,31 @@ public static long offsetFromFileName(String fileName) {
+     * @return a File instance for producer snapshot.
+     */
+    public static File producerSnapshotFile(File logDir, long offset) {
+        return new File(logDir, filenamePrefixFromOffset(offset) + LogFileUtils.PRODUCER_SNAPSHOT_FILE_SUFFIX);
```
`LogFileUtils.` is redundant since it's the same file?

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogFileUtils.java:
```diff
@@ -32,4 +37,31 @@ public static long offsetFromFileName(String fileName) {
+     *
+     * @param logDir parent directory for the snapshot file
+     * @param offset offset number for the snapshot file
+     * @return a File instance for producer snapshot.
```
This documentation is different from the `UnifiedLog` one. In particular, the details for `offset` are important (and now missing).

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogFileUtils.java:
```diff
@@ -16,7 +16,12 @@
+
 public final class LogFileUtils {
+    public static final String PRODUCER_SNAPSHOT_FILE_SUFFIX = ".snapshot";
+    public static final String DELETED_FILE_SUFFIX = ".deleted";
```
Worth adding the comment from `LocalLog`:

`/** a file that is scheduled to be deleted */`

ijuma commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -680,19 +680,23 @@ class UnifiedLog(@volatile var logStartOffset: Long,
   }
 
   private[log] def activeProducersWithLastSequence: Map[Long, Int] = lock synchronized {
```
Is it right that this is only used in tests? If so, can we move this to `LogTestUtils`? Then we don't have to worry about the `toMap` copy. Alternatively, just change the return type to `scala.collection.Map` and remove the unnecessary `toMap` at the end.

satishd replied:
Right, it is used only in tests. But we should not move to `LogTestUtils` as this method should take a lock while building the results. It is better to keep these semantics and leave this method here even though it is used only in tests. This will keep the method contract clean and thread safe.

ijuma replied:
Makes sense about keeping the implementation in `UnifiedLog`. I still think we should simply return `collection.Map` and avoid the extra copy.

ijuma commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -680,19 +680,23 @@ class UnifiedLog(@volatile var logStartOffset: Long,
+    result.toMap
   }
 
   private[log] def lastRecordsOfActiveProducers: Map[Long, LastRecord] = lock synchronized {
```
We should return `scala.collection.Map` here and remove the `toMap` at the end of the implementation.

ijuma reviewed (commented):
Thanks for the update. A few comments below. I mostly went through the previous comments. Since so much has changed, I need to go through `ProducerStateManager` one more time.

satishd committed:
Addressed review comments

ijuma reviewed (approved):
Once https://github.com/apache/kafka/pull/13040#discussion_r1102828124 is done, LGTM. Thanks!

satishd committed:
Addressed reveiw comments
Replaced file.isFile with Files.isRegularFile(path) as suhhested in the review.

ijuma commented:
Looks like the last commit caused a number of test failures.

satishd committed:
Fixed ProducerStateManager.isSnapshotFile to use toString instead of directly checking with Paths as the check is with a suffix but not a complete path.

ijuma commented:
@satishd the latest CI run is green.

satishd commented:
All the tests are passed, merging to trunk.

satishd commented:
Sorry for missed adding the reviewers section in the merge commit.
